Vectorize RMSNorm CUDA kernel

bbeckca opened the pull request:
## Summary:
•	Vectorize writes when in/out phases align; scalar reads for weight when needed
•	Use an optional fp16 shared-mem cache (avoids second read)
•	Preserve numerics on odd/strided shapes by keeping scalar fallback
•	Match unfused launch/reduction order

## Test Result

**Accuracy **
```

lm_eval   --model vllm   --model_args "pretrained=Qwen/Qwen3-30B-A3B-FP8,max_model_len=32768,enforce_eager=True"   --trust_remote_code   --tasks gsm8k   --num_fewshot 5   --batch_size auto

# Before
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8279|±  |0.0104|
|     |       |strict-match    |     5|exact_match|↑  |0.8878|±  |0.0087|

# After
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8218|±  |0.0105|
|     |       |strict-match    |     5|exact_match|↑  |0.8939|±  |0.0085|

```

```
(vllm-env) bbeckca@computeinstance-e00z6c2em7n8zqhbdr:~/vllm$ pytest tests/kernels/core/test_layernorm.py
======================= test session starts =======================
platform linux -- Python 3.12.3, pytest-8.4.2, pluggy-1.6.0
rootdir: /home/bbeckca/vllm
configfile: pyproject.toml
plugins: anyio-4.10.0
collected 1584 items                                              

tests/kernels/core/test_layernorm.py ...................... [  1%]
........................................................... [  5%]
........................................................... [  8%]
........................................................... [ 12%]
........................................................... [ 16%]
........................................................... [ 20%]
........................................................... [ 23%]
........................................................... [ 27%]
........................................................... [ 31%]
........................................................... [ 34%]
........................................................... [ 38%]
........................................................... [ 42%]
........................................................... [ 46%]
........................................................... [ 49%]
........................................................... [ 53%]
........................................................... [ 57%]
........................................................... [ 60%]
........................................................... [ 64%]
........................................................... [ 68%]
........................................................... [ 72%]
........................................................... [ 75%]
........................................................... [ 79%]
........................................................... [ 83%]
........................................................... [ 87%]
........................................................... [ 90%]
........................................................... [ 94%]
........................................................... [ 98%]
............................                                [100%]

================ 1584 passed in 375.51s (0:06:15) =================
```

**Performance**

```
python benchmark_layernorm.py

| Tokens | Hidden | Dtype    | Old (µs) | New (µs) | Change (%) |
|-------:|-------:|:---------|---------:|---------:|-----------:|
|      2 |   1024 | bfloat16 |   13.544 |   13.444 |      -0.7% |
|      2 |   1024 | half     |   13.628 |   13.567 |      -0.4% |
|      2 |   8192 | bfloat16 |   13.673 |   13.801 |      +0.9% |
|      2 |   8192 | half     |   13.779 |   13.741 |      -0.3% |
|      4 |   1024 | bfloat16 |   13.732 |   13.662 |      -0.5% |
|      4 |   1024 | half     |   13.692 |   13.399 |      -2.1% |
|      4 |   8192 | bfloat16 |   13.646 |   14.323 |      +5.0% |
|      4 |   8192 | half     |   14.483 |   13.464 |      -7.0% |
|      8 |   1024 | bfloat16 |   13.664 |   13.595 |      -0.5% |
|      8 |   1024 | half     |   13.687 |   13.914 |      +1.7% |
|      8 |   8192 | bfloat16 |   13.472 |   13.629 |      +1.2% |
|      8 |   8192 | half     |   13.235 |   13.314 |      +0.6% |
|     16 |   1024 | bfloat16 |   13.451 |   13.264 |      -1.4% |
|     16 |   1024 | half     |   13.718 |   13.357 |      -2.6% |
|     16 |   8192 | bfloat16 |   13.207 |   13.332 |      +0.9% |
|     16 |   8192 | half     |   13.690 |   13.673 |      -0.1% |
|     32 |   1024 | bfloat16 |   13.998 |   13.560 |      -3.1% |
|     32 |   1024 | half     |   13.687 |   13.359 |      -2.4% |
|     32 |   8192 | bfloat16 |   13.671 |   13.661 |      -0.1% |
|     32 |   8192 | half     |   13.401 |   13.362 |      -0.3% |
|     64 |   1024 | bfloat16 |   13.587 |   13.600 |      +0.1% |
|     64 |   1024 | half     |   13.691 |   13.453 |      -1.7% |
|     64 |   8192 | bfloat16 |   13.571 |   13.648 |      +0.6% |
|     64 |   8192 | half     |   13.851 |   13.505 |      -2.5% |
|    128 |   1024 | bfloat16 |   13.882 |   13.577 |      -2.2% |
|    128 |   1024 | half     |   13.325 |   13.681 |      +2.7% |
|    128 |   8192 | bfloat16 |   13.960 |   13.663 |      -2.1% |
|    128 |   8192 | half     |   13.844 |   13.897 |      +0.4% |
|    256 |   1024 | bfloat16 |   13.469 |   13.655 |      +1.4% |
|    256 |   1024 | half     |   14.165 |   13.392 |      -5.5% |
|    256 |   8192 | bfloat16 |   13.734 |   13.834 |      +0.7% |
|    256 |   8192 | half     |   13.432 |   13.725 |      +2.2% |
|    512 |   1024 | bfloat16 |   13.612 |   13.556 |      -0.4% |
|    512 |   1024 | half     |   13.954 |   13.409 |      -3.9% |
|    512 |   8192 | bfloat16 |   13.658 |   13.703 |      +0.3% |
|    512 |   8192 | half     |   13.554 |   13.532 |      -0.2% |
|   1024 |   1024 | bfloat16 |   13.760 |   13.615 |      -1.1% |
|   1024 |   1024 | half     |   13.868 |   13.805 |      -0.5% |
|   1024 |   8192 | bfloat16 |   24.042 |   16.055 |     -33.2% |
|   1024 |   8192 | half     |   24.103 |   15.766 |     -34.6% |
|   2048 |   1024 | bfloat16 |   13.683 |   13.656 |      -0.2% |
|   2048 |   1024 | half     |   13.445 |   13.720 |      +2.0% |
|   2048 |   8192 | bfloat16 |   60.042 |   43.753 |     -27.1% |
|   2048 |   8192 | half     |   59.763 |   43.622 |     -27.0% |
|   4096 |   1024 | bfloat16 |   23.686 |   13.645 |     -42.4% |
|   4096 |   1024 | half     |   23.456 |   13.514 |     -42.4% |
|   4096 |   8192 | bfloat16 |  114.548 |   81.739 |     -28.6% |
|   4096 |   8192 | half     |  114.347 |   81.835 |     -28.4% |
|   8192 |   1024 | bfloat16 |   46.955 |   20.395 |     -56.6% |
|   8192 |   1024 | half     |   46.964 |   19.760 |     -57.9% |
|   8192 |   8192 | bfloat16 |  220.422 |  156.080 |     -29.2% |
|   8192 |   8192 | half     |  220.369 |  156.161 |     -29.1% |
|  16384 |   1024 | bfloat16 |  100.117 |   51.102 |     -49.0% |
|  16384 |   1024 | half     |  100.163 |   41.369 |     -58.7% |
|  16384 |   8192 | bfloat16 |  433.641 |  303.721 |     -30.0% |
|  16384 |   8192 | half     |  432.265 |  303.430 |     -29.8% |
```

Differential Revision: D79969610

bbeckca commented on csrc/layernorm_kernels.cu:
```diff
@@ -12,35 +13,254 @@
 
-// TODO(woosuk): Further optimize this kernel.
+// ---------- helpers ----------
+template <typename T>
+__device__ __forceinline__ T warp_sum(T v) {
+#ifdef __HIP_PLATFORM_AMD__
+  const unsigned long long m = 0xffffffffffffffffull;  // HIP needs 64-bit mask
+#else
+  const unsigned m = 0xffffffffu;  // CUDA 32-bit mask
+#endif
+  // Always reduce over 32 lanes to match downstream logic.
+  constexpr int kWidth = 32;
+  v += __shfl_down_sync(m, v, 16, kWidth);
+  v += __shfl_down_sync(m, v, 8, kWidth);
+  v += __shfl_down_sync(m, v, 4, kWidth);
+  v += __shfl_down_sync(m, v, 2, kWidth);
+  v += __shfl_down_sync(m, v, 1, kWidth);
+  return v;
+}
```
Note: `warp_sum` and `ln_block_threads_unified` are the same for rms_norm_kernel and rms_norm_static_fp8_quant. Will move to shared helper after aligning on high level approach.

yewentao256 reviewed (commented):
Thanks for the work!

Could we also test for end to end performance & accuracy?

For accuracy: `lm_eval ...`

For E2E performance (optional, nice to have): `vllm bench ...`

bbeckca commented on csrc/layernorm_quant_kernels.cu:
```diff
@@ -76,7 +105,7 @@ fused_add_rms_norm_static_fp8_quant_kernel(
 
   const int vec_hidden_size = hidden_size / width;
-  const int vec_input_stride = input_stride / width;
+  const int64_t vec_input_stride = input_stride / width;
```
Turns out this change from `int` to `int64t` wasn't needed, so I've removed it entirely.

Commit:
Vectorize RMSNorm CUDA kernel (#22602)

Summary:

What: Make RMSNorm faster by reading data in bigger aligned chunks, caching fp16 rows in shared memory, making the FP8-quant version work with strided inputs, and using the same launch settings as the unfused path.

Why: Cut global memory traffic using aligned vector inputs/outputs and shared-mem reuse (avoids second read), make the FP8 path safe for strided inputs, and preserve numerics by matching the unfused reduction/launch order.

Test Plan:
1) Run tests

```
[benjibeck@devvm8203.cco0 /data/users/benjibeck/fbsource/fbcode/vllm (1043a27694)]$ buck2 test :test_kernels_layernorm
Buck UI: https://www.internalfb.com/buck2/054ebad3-ad92-4676-a4d2-3bf43e44f31a
Test UI: https://www.internalfb.com/intern/testinfra/testrun/10414574240710255
Network: Up: 152MiB  Down: 2.9GiB  (reSessionID-14af330c-26bf-41d5-87b0-5775bf7d6f8a)
Loading targets.   Remaining     0/7                                                                                           150 dirs read, 69 targets declared
Analyzing targets. Remaining     0/32                                                                                          772 actions, 819 artifacts declared
Executing actions. Remaining     0/245                                                                                         48.3s exec time total
Command: test.     Finished 1 local, 14 remote, 131 cache (90% hit)                                                            45.2s exec time cached (93%)
Time elapsed: 4:53.4s
Tests finished: Pass 3169. Fail 0. Fatal 0. Skip 0. Build failure 0
```

2) Run benchmark
```
buck run :benchmark_layernorm -- --num-tokens 16384 --hidden-size 1024 --dtype half --num-iters 500
```

Performance (selected, non-strided / HND)

| T | H | dtype | baseline (µs) | current (µs) | Δ |
| 4096  | 1024 | fp16 | 24.592  | 16.552  | -32.7% |
| 16384 | 1024 | fp16 | 106.699 | 42.739  | -60.0% |
| 4096  | 8192 | fp16 | 118.566 | 97.059  | -18.1% |
| 16384 | 8192 | fp16 | 450.738 | 356.125 | -21.0% |
| 4096  | 1024 | bf16 | 24.743  | 16.683  | -32.6% |
| 16384 | 1024 | bf16 | 107.009 | 56.946  | -46.8% |
| 4096  | 8192 | bf16 | 119.293 | 96.774  | -18.9% |
| 16384 | 8192 | bf16 | 451.181 | 357.799 | -20.7% |

Strided (NHD) overhead (current kernel) — penalty vs. HND (same T/H/dtype):
- 4096×1024 fp16: 1.39× (22.983 / 16.552)
- 16384×1024 fp16: 2.13× (90.995 / 42.739)
- 4096×8192 fp16: 1.93× (186.931 / 97.059)

Rollback Plan:

Differential Revision: D79969610

Signed-off-by: Benji Beck <benjibeck@meta.com>

Commit:
Fix pre commit

Signed-off-by: Benji Beck <benjibeck@meta.com>

bbeckca commented:
Update: Performed e2e benchmark for latency using x2 H100 and observed ~4.8% reduction with vectorized rms_norm_kernel without shared memory. I'm planning to benchmark further with other models/configurations to understand this better. Let me know if you folks have any thoughts/preferences.

```
VLLM_USE_V1=1 buck run //vllm:benchmark_latency --   --model facebook/opt-125m   --input-len 32 --output-len 128   --dtype float16 --batch-size 128 --enforce-eager

baseline
Avg latency: 0.6731001150328666 seconds
10% percentile latency: 0.6315004411269911 seconds
25% percentile latency: 0.6374467767454917 seconds
50% percentile latency: 0.6609808525245171 seconds
75% percentile latency: 0.6663343247637386 seconds
90% percentile latency: 0.6783564629906322 seconds
99% percentile latency: 1.0606678462994754 seconds

vectorized
Avg latency: 0.6741364902312247 seconds
10% percentile latency: 0.6488532381190453 seconds
25% percentile latency: 0.6496474424784537 seconds
50% percentile latency: 0.6550397415121552 seconds
75% percentile latency: 0.6627388229971984 seconds
90% percentile latency: 0.666680197842652 seconds
99% percentile latency: 1.039106485189987 seconds

vectorized (without shared memory)
Avg latency: 0.6409522683670124 seconds
10% percentile latency: 0.6146808989578858 seconds
25% percentile latency: 0.6191042929858668 seconds
50% percentile latency: 0.6249049424950499 seconds
75% percentile latency: 0.6411205744952895 seconds
90% percentile latency: 0.6505832085269503 seconds
99% percentile latency: 0.8902701860625533 seconds
```

bbeckca commented:
Ran quick E2E latency on a few models (bs=128, in=32, out=128). Overall: small win on Llama-3.2, neutral on Qwen fp16, and Gemma-2-9B works better if we skip the shared-mem cached row. I can rerun with further settings or add other models if there's any preferences.

@yewentao256  Wondering if we should consider adding a knob to disable the cached-row path where it regresses for some models?

bbeckca commented:
> Ran quick E2E latency on a few models (bs=128, in=32, out=128). Overall: small win on Llama-3.2, neutral on Qwen fp16, and Gemma-2-9B works better if we skip the shared-mem cached row. I can rerun with further settings or add other models if there's any preferences.
> 
> @yewentao256 Wondering if we should consider adding a knob to disable the cached-row path where it regresses for some models?
> 
> Model	dtype	baseline (s)	vectorized (s)	Δ vs base	no-shared-mem (s)	Δ vs base
> Llama-3.2-3B-Instruct	fp16	1.2895	1.2698	**−1.5%**	1.2863	−0.3%
> Llama-3.2-3B-Instruct	bf16	1.2772	1.2663	**−0.9%**	1.2892	+0.9%
> Qwen2-7B	fp16	1.7673	1.7692	+0.1%	1.7679	+0.0%
> Qwen2-7B	bf16	1.7307	1.7143	**−0.9%**	1.6929	**−2.2%**
> Gemma-2-9B	bf16	4.1187	4.2222	**+2.5%**	4.0485	**−1.7%**

cc @mgoin @WoosukKwon for any additional thoughts.

yewentao256 reviewed (commented):
Sorry for the late response, there are some problems with my github notifications but now fixed.
The overall work I think is good, could you
1. test the accuracy using `lm_eval`?
2. benchmark the kernel directly?
One example could be seen here: https://github.com/vllm-project/vllm/pull/22036

Commit:
Restore benchmark_layernorm.py to match upstream/main

Signed-off-by: Benji Beck <benjibeck@meta.com>

bbeckca commented:
Sorry for the noise! Had some branch hygiene issue with new setup. Cleaned it up but apologies for any inconveniences.

bbeckca commented:
> Sorry for the late response, there are some problems with my github notifications but now fixed. The overall work I think is good, could you
> 
> 1. test the accuracy using `lm_eval`?
> 2. benchmark the kernel directly?
>    One example could be seen here: #22036

Thanks for your patience. I’ve updated the test plan following your guidance. There are some regressions at very small sizes, the kernel shows significant performance gains for 1k+ tokens. Could you take another look?

yewentao256 commented on csrc/layernorm_kernels.cu:
```diff
@@ -12,35 +13,232 @@
+  if (vec_store_ok) {
+    ScalarMulNormWeight<scalar_t> sca_op{weight, out_row, inv_rms};
+
+    const auto addr = reinterpret_cast<uintptr_t>(vin);
+    const int mis = addr & (WIDTH - 1);
+    const int prefix =
+        mis ? (WIDTH - mis) / static_cast<int>(sizeof(scalar_t)) : 0;
```
I think for `vectorize_with_alignment` we have the prefix handle, could we remove this?

yewentao256 commented on csrc/layernorm_kernels.cu:
```diff
@@ -355,18 +560,32 @@ void rms_norm(torch::Tensor& out,     // [..., hidden_size]
   const at::cuda::OptionalCUDAGuard device_guard(device_of(input));
   const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+  // Optional per-block row cache in dynamic shared memory.
+  // If enabled (FP16 and HS <= 4096), the kernel copies the row to smem once
+  // and reuses it on the second pass to cut a global read. If shmem_bytes == 0,
+  // the kernel takes the non-cached path
+  size_t shmem_bytes = 0;
+  int smem_elems = 0;
+  if (input.scalar_type() == at::kHalf && hidden_size <= 4096) {
+    shmem_bytes = static_cast<size_t>(hidden_size) * sizeof(at::Half);
+    smem_elems = hidden_size;  // flag to kernel that shmem was provisioned
+  }
+
```
Do we have the quantitative benefits results for introducing this smem change?

yewentao256 commented on csrc/layernorm_quant_kernels.cu:
```diff
@@ -20,40 +20,67 @@
 namespace vllm {
 
-// TODO(woosuk): Further optimize this kernel.
+template <typename T>
+__device__ __forceinline__ T warp_sum(T v) {
+#ifdef __HIP_PLATFORM_AMD__
+  const unsigned long long m = 0xffffffffffffffffull;  // HIP needs 64-bit mask
+#else
+  const unsigned m = 0xffffffffu;
+#endif
+  constexpr int kWidth = 32;  // keep reduction over 32 lanes everywhere
+  v += __shfl_down_sync(m, v, 16, kWidth);
+  v += __shfl_down_sync(m, v, 8, kWidth);
+  v += __shfl_down_sync(m, v, 4, kWidth);
+  v += __shfl_down_sync(m, v, 2, kWidth);
+  v += __shfl_down_sync(m, v, 1, kWidth);
+  return v;
+}
```
Seems identical with previous code, is there any way we can reuse it instead of copying it twice?

yewentao256 commented on csrc/layernorm_quant_kernels.cu:
```diff
@@ -169,20 +196,37 @@ fused_add_rms_norm_static_fp8_quant_kernel(
+  threads = std::min(1024, std::max(128, ((threads + 31) / 32) * 32));
+  return threads;
+}
+
```
reuse code here to avoid duplicate code
And please also add comments why we need this function?

yewentao256 commented on csrc/layernorm_quant_kernels.cu:
```diff
@@ -191,10 +235,11 @@ void rms_norm_static_fp8_quant(torch::Tensor& out,     // [..., hidden_size]
+                      num_tokens, hidden_size);
             });
       });
+  // TORCH_CUDA_KERNEL_LAUNCH_CHECK();
```
Remove unused code comment

yewentao256 reviewed (commented):
Thanks for the work!

Commit:
Move to warp_sum and ln_block_threads_unified to layernorm_utils.cuh

Signed-off-by: Benji Beck <benjibeck@meta.com>

bbeckca commented:
@yewentao256 Heads up, I’m hitting some local env issues so my RMSNorm revisions are taking longer than expected. Working through it and will follow up once it's ready.

bbeckca commented:
Hey @yewentao256, thanks for the prior reviews! I plan to close this diff and restart from a clean slate since recurring build issues made it messy. I’ll share the new diff once it’s ready. Just checking if you’re still on board with making these changes? Feel free to let me know if any preferences.

yewentao256 commented:
@bbeckca Sure, feel free to have a new PR
